dfa_embeddings/dfa_builder.py

```diff
@@ -0,0 +1,227 @@
+# import ring
+import random
+import numpy as np
+
+import dgl
+import torch
+import networkx as nx
+from copy import deepcopy
+from dfa import DFA
+
+# feature_inds = {"rejecting": -1, "accepting": -2, "temp": -3, "normal": -4, "init": -5, "AND": -6, "OR": -7, "distance_normalized": -8}
+# feature_inds = {"rejecting": -1, "accepting": -2, "temp": -3, "normal": -4, "init": -5, "AND": -6, "distance_normalized": -7}
+feature_inds = {"rejecting": -1, "accepting": -2, "temp": -3, "normal": -4, "init": -5, "AND": -6}
+edge_types = {k:v for (v, k) in enumerate(["self", "normal-to-temp", "temp-to-normal", "AND"])}
+# edge_types = {k:v for (v, k) in enumerate(["self", "normal-to-temp", "temp-to-normal", "AND", "OR"])}
+
+"""
+A class that can take an DFA formula and generate the Abstract Syntax Tree (DFA) of it. This
+code can generate trees in either Networkx or DGL formats. And uses caching to remember recently
+generated trees.
+"""
```

Comment:
Doc strings go inside class / function definitions in python:

https://peps.python.org/pep-0257/